### PR TITLE
cluster: support external timestamp

### LINKED ISSUE #5637: support external timestamp
### Describe your feature request related problem

support external timestamp

### Describe the feature you'd like

support external timestamp getter and setter

### PR DESCRIPTION
**Pull request by lhy1024**
Signed-off-by: lhy1024 <admin@liudos.us>

### What problem does this PR solve?

Issue Number: close #5637, ref https://github.com/pingcap/tidb/issues/38274

### What is changed and how does it work?

This pr supports external timestamp getter and setter without client batch.

The new external timestamp should be larger than now and it should be less or equal to global ts.

### Check List

Tests

- Unit test
- Integration test
- Manual test (add detailed scripts or steps below)

Disable
```
SQL statistics:
    queries performed:
        read:                            329840
        write:                           0
        other:                           47120
        total:                           376960
    transactions:                        23560  (392.59 per sec.)
    queries:                             376960 (6281.42 per sec.)
    ignored errors:                      0      (0.00 per sec.)
    reconnects:                          0      (0.00 per sec.)

General statistics:
    total time:                          60.0113s
    total number of events:              23560

Latency (ms):
         min:                                    6.66
         avg:                                   10.18
         max:                                   24.88
         95th percentile:                       11.65
         sum:                               239957.27

Threads fairness:
    events (avg/stddev):           5890.0000/1.22
    execution time (avg/stddev):   59.9893/0.01
```
Enable：
```
SQL statistics:
    queries performed:
        read:                            325388
        write:                           0
        other:                           46484
        total:                           371872
    transactions:                        23242  (387.31 per sec.)
    queries:                             371872 (6196.92 per sec.)
    ignored errors:                      0      (0.00 per sec.)
    reconnects:                          0      (0.00 per sec.)

General statistics:
    total time:                          60.0083s
    total number of events:              23242

Latency (ms):
         min:                                    6.72
         avg:                                   10.32
         max:                                   23.19
         95th percentile:                       11.87
         sum:                               239958.61

Threads fairness:
    events (avg/stddev):           5810.5000/0.50
    execution time (avg/stddev):   59.9897/0.00
```
### Release note

```release-note
None.
```

### CODE REVIEW

File: client/client.go [hunk 1]
```diff
@@ -1885,6 +1890,30 @@ func (c *client) WatchGlobalConfig(ctx context.Context) (chan []GlobalConfigItem
 	return globalConfigWatcherCh, err
 }
 
+func (c *client) GetExternalTimestamp(ctx context.Context) (uint64, error) {
+	resp, err := c.getClient().GetExternalTimestamp(ctx, &pdpb.GetExternalTimestampRequest{})
```
Comment (YangKeao) [inline client/client.go:1894]: Should it add header?

```go
&pdpb.GetExternalTimestampRequest{
	Header: c.requestHeader(),
}
```

As the handler will validate the clusterId

**Comment by JmPotato** [comment]
Could we split this PR into several ones? For example:

* Storage API implementation
* HTTP/Client API implementation
* ...

**Review by YangKeao — APPROVED** [review]
LGTM

**Comment by lhy1024** [comment]
> Could we split this PR into several ones? For example:
> 
> * Storage API implementation
> * HTTP/Client API implementation
> * ...

I don't think it's necessary because although the code looks a lot, the logic is simple.

File: server/api/external_timestamp.go [hunk 2]
```diff
@@ -0,0 +1,52 @@
...
+}
+
+// @Tags     external_timestamp
+// @Summary  Get external timestamp.
+// @Produce  json
+// @Success  200  {array}   ExternalTimestamp
+// @Failure  500  {string}  string  "PD server failed to proceed the request."
+// @Router   /external-timestamp [get]
+func (h *externalTimestampHandler) GetExternalTS(w http.ResponseWriter, r *http.Request) {
+	value := h.svr.GetExternalTS()
```
Comment (nolouch) [inline server/api/external_timestamp.go:48]: why need a HTTP API?

Comment (lhy1024) [reply server/api/external_timestamp.go:48]: only to provide debug

Comment (nolouch) [reply server/api/external_timestamp.go:48]: should remove it if no needed.

File: client/client.go [hunk 3]
```diff
@@ -1885,6 +1890,35 @@ func (c *client) WatchGlobalConfig(ctx context.Context) (chan []GlobalConfigItem
 	return globalConfigWatcherCh, err
 }
 
+func (c *client) GetExternalTimestamp(ctx context.Context) (uint64, error) {
+	resp, err := c.getClient().GetExternalTimestamp(ctx, &pdpb.GetExternalTimestampRequest{
```
Comment (nolouch) [inline client/client.go:1894]: not supported batch here?

File: server/storage/endpoint/key_path.go [hunk 4]
```diff
@@ -106,11 +107,16 @@ func gcSafePointServicePath(serviceID string) string {
 	return path.Join(gcSafePointPath(), "service", serviceID)
 }
 
-// MinResolvedTSPath returns the min resolved ts path
+// MinResolvedTSPath returns the min resolved ts path.
 func MinResolvedTSPath() string {
 	return path.Join(clusterPath, minResolvedTS)
 }
 
+// ExternalTimestampPath returns the external timestamp path.
```
Comment (bufferflies) [inline server/storage/endpoint/key_path.go:115]: how about using constant string ?

File: server/storage/storage.go [hunk 5]
```diff
@@ -39,6 +39,7 @@ type Storage interface {
 	endpoint.ReplicationStatusStorage
 	endpoint.GCSafePointStorage
 	endpoint.MinResolvedTSStorage
+	endpoint.ExternalTSStorage
```
Comment (bufferflies) [inline server/storage/storage.go:42]: why add this and where is it used?

File: client/client.go (hunk 3 above)
Comment (lhy1024) [reply client/client.go:1894]: It not supported batch in this pr

File: server/storage/storage.go (hunk 5 above)
Comment (lhy1024) [reply server/storage/storage.go:42]: persist external ts

File: server/storage/endpoint/external_timestamp.go [hunk 6]
```diff
@@ -0,0 +1,54 @@
...
+
+var _ ExternalTSStorage = (*StorageEndpoint)(nil)
+
+// LoadExternalTS loads the external timestamp from storage.
+func (se *StorageEndpoint) LoadExternalTS() (uint64, error) {
+	value, err := se.Load(ExternalTimestampPath())
+	if err != nil || value == "" {
+		return 0, err
+	}
+	timestamp, err := strconv.ParseUint(value, 16, 64)
```
Comment (bufferflies) [inline server/storage/endpoint/external_timestamp.go:43]: why use 16bit?

Comment (lhy1024) [reply server/storage/endpoint/external_timestamp.go:43]: I found that gc-safe-point and min-resolved-ts are saved directly with uint64, but tso is converted to bytes for saving. Does external-ts need to be in the same format as tso, or any one of them? cc @JmPotato

File: server/storage/endpoint/key_path.go (hunk 4 above)
Comment (lhy1024) [reply server/storage/endpoint/key_path.go:141]: Just keep the same with other paths

File: server/storage/endpoint/external_timestamp.go (hunk 6 above)
Comment (nolouch) [reply server/storage/endpoint/external_timestamp.go:43]: external-ts set from cdc, so it should consist with tso,

Comment (lhy1024) [reply server/storage/endpoint/external_timestamp.go:43]: Do they also need to be stored in the same format in etcd?

**Review by nolouch — APPROVED** [review]

**Review by HunDunDM — APPROVED** [review]

**Review by HunDunDM — APPROVED** [review]

**Pull request merged**